src/librustc_typeck/check/_match.rs

```diff
@@ -595,73 +600,24 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
-        // Then if the pattern matches by reference, we want to match
-        // `f()[0]` as a lexpr, so we can't allow it to be
-        // coerced. But if the pattern matches by value, `f()[0]` is
-        // still syntactically a lexpr, but we *do* want to allow
-        // coercions.
-        //
-        // However, *likely* we are ok with allowing coercions to
-        // happen if there are no explicit ref mut patterns - all
-        // implicit ref mut patterns must occur behind a reference, so
-        // they will have the "correct" variance and lifetime.
-        //
-        // This does mean that the following pattern would be legal:
-        //
-        // ```
-        // struct Foo(Bar);
-        // struct Bar(u32);
-        // impl Deref for Foo {
-        //     type Target = Bar;
-        //     fn deref(&self) -> &Bar { &self.0 }
-        // }
-        // impl DerefMut for Foo {
-        //     fn deref_mut(&mut self) -> &mut Bar { &mut self.0 }
-        // }
-        // fn foo(x: &mut Foo) {
-        //     {
-        //         let Bar(z): &mut Bar = x;
-        //         *z = 42;
-        //     }
-        //     assert_eq!(foo.0.0, 42);
-        // }
-        // ```
-        //
-        // FIXME(tschottdorf): don't call contains_explicit_ref_binding, which
-        // is problematic as the HIR is being scraped, but ref bindings may be
-        // implicit after #42640. We need to make sure that pat_adjustments
-        // (once introduced) is populated by the time we get here.
-        //
-        // See #44848.
-        let contains_ref_bindings = arms.iter()
-                                        .filter_map(|a| a.contains_explicit_ref_binding())
-                                        .max_by_key(|m| match *m {
-                                            hir::MutMutable => 1,
-                                            hir::MutImmutable => 0,
-                                        });
-        let discrim_ty;
-        if let Some(m) = contains_ref_bindings {
-            discrim_ty = self.check_expr_with_needs(discrim, Needs::maybe_mut_place(m));
+        use hir::MatchSource::*;
+        let (source_if, if_no_else, if_desugar) = match match_src {
+            IfDesugar { contains_else_clause } => (true, !contains_else_clause, true),
+            IfLetDesugar { contains_else_clause } => (true, !contains_else_clause, false),
+            _ => (false, false, false),
+        };
+
+        // Type check the descriminant and get its type.
+        let discrim_ty = if if_desugar {
+            // Here we want to ensure:
+            //
+            // 1. That default match bindings are *not* accepted in the condition of an
+            //    `if` expression. E.g. given `fn foo() -> &bool;` we reject `if foo() { .. }`.
+            //
+            // 2. By expecting `bool` for `expr` we get nice diagnostics for e.g. `if x = y { .. }`.
+            self.check_expr_has_type_or_error(discrim, self.tcx.types.bool)
```

Comment:
Do not use desugaring information to affect the semantics, *only* adjust diagnostics.
This should be handled by generating a type-ascription `cond: bool` expression, for `if cond {...}`.

Reply:
(As discussed on Discord in private conversation...)

So I tried both `DropTemps($cond): bool` and `DropTemps($cond: bool)` and both solutions has the problem that `$expr: $type` does not take coercions into account which means that `loop { if break () {} }` which compiles today stops compiling because it requires a coercion `! -> bool`. [As we know](https://github.com/rust-lang/rust/issues/23416), type ascription is sadly broken (it should allow that coercion to happen but it requires an exact match). Moreover, both solutions will likely require more hacks to improve diagnostics again which regressed.

I see three options to solve this in a different manner:
1. Use `$cond as bool` -- this seems more complex than the hack.
2. Embed `Option<PrimTy>` into `DropTemps` and handle the coercion logic in here; this is probably still more complicated than the hack.
3. Remove the hack and accept the semantic change. (requires T-Lang sign-off + FCP)

As it stands right now, the hack is fairly simple and isn't that hard to remove later. The hack is also well tested in the PR so we can be sure there are no correctness issues. However, it does make `if` into something other than a pure desugaring. That's inelegant, but in my view it works as a stop-gap.

I would like to consider option 3. eventually because it improves usability but this is in my view out of scope for this PR.

cc @nikomatsakis -- @eddyb wanted you to sign off on the hack.

Reply:
> Do not use desugaring information to affect the semantics, only adjust diagnostics.

I thought only `Span` expansion info should not be used to affect semantics. If `MatchSource` should only be used for diagnostics, would it make sense to move all of that info into `ExpansionInfo`?

I also like option 3, is there already an issue about that?

Reply:
@oli-obk There's no issue about that to my knowledge as there wasn't a implementation-driven need before. But now that `if` is almost a pure desugaring to `match` there's more justification for doing it (aside from the usability benefits).

One drawback to 3. is that we no longer have `ExpectHasType(bool)` to trigger good diagnostics for `if x = y {`. If we care deeply about that diagnostic trigger we might need to generate a skeleton from the pattern so that we get the expectation back.

For now however, I think we shouldn't try to do too much in this PR. I've left a `FIXME` addressed at rust-lang/lang-team to think about the issue later.

Reply:
In what way is using `$cond as bool` more complex than the hack?

Reply:
@nikomatsakis My thinking is casts have more complex machinery so `$cond as bool` has more logic. The use itself is probably not that complicated but then again the diagnostics need to be tweaked to account for the cast in-between which is probably more complicated than the hack.

Reply:
To be clear, I'm happy to try out `$cond as bool` if that is desired.

Reply:
(Not questioning that it is, it's just not entirely obvious to me)

Reply:
@nikomatsakis Same! I'm not sure the hack will be best once I actually implement `let_chains`. It may turn out that `$cond as bool` is cleanest when do that but it should be fairly easy to switch if that is the case. =)